test/unitTests/build/reportGenerator.test.ts

```diff
@@ -0,0 +1,167 @@
+
+    let eventStream: EventStream;
+    let testEventBus: TestEventBus;
+
+    let sinon: SinonSandbox;
+    let stubSafelyReadJsonFile: SinonStub;
+    let stubFsExistsSync: SinonStub;
+    let stubFsReadFileSync: SinonStub;
+
+    let diagnosticSet: {
+        [key: string]: any;
+    } = {};
+    const fakedDiagnosticController = <DiagnosticController>{
+        reset: () => {
+            diagnosticSet = {};
+        },
+        setDiagnostic: (uri: Uri, diagnostics: Diagnostic[]) => {
+            diagnosticSet[uri.fsPath] = {
+                uri,
+                diagnostics
+            };
+        }
+    };
+
+    const expectedInfoDiagnostic = new Diagnostic(new Range(0, 0, 0, 0), `Missing required attribute: 'author'. Add the current author's GitHub ID.`, vscode.DiagnosticSeverity.Hint);
+    const expectedWarningDiagnostic = new Diagnostic(new Range(0, 0, 0, 0), `Missing required attribute: 'author'. Add the current author's GitHub ID.`, vscode.DiagnosticSeverity.Warning);
+    const expectedErrorDiagnostic = new Diagnostic(new Range(0, 0, 0, 0), `Missing required attribute: 'author'. Add the current author's GitHub ID.`, vscode.DiagnosticSeverity.Error);
+    expectedInfoDiagnostic.code = 'author-missing';
+    expectedInfoDiagnostic.source = 'Docs Validation';
+    expectedWarningDiagnostic.code = 'author-missing';
+    expectedWarningDiagnostic.source = 'Docs Validation';
+    expectedErrorDiagnostic.code = 'author-missing';
+    expectedErrorDiagnostic.source = 'Docs Validation';
+    const expectedFile1Uri = Uri.file(`${testRepositoryPath}/sourceFolder1/index.md`);
+    const expectedFile2Uri = Uri.file(`${testRepositoryPath}/sourceFolder2/index.md`);
+    expectedFile1Uri.fsPath;
+    expectedFile2Uri.fsPath;
+
+    before(() => {
+        eventStream = new EventStream();
+        testEventBus = new TestEventBus(eventStream);
+        sinon = createSandbox();
+
+        stubSafelyReadJsonFile = sinon.stub(utils, "safelyReadJsonFile");
+        stubFsExistsSync = sinon.stub(fs, "existsSync");
+        stubFsReadFileSync = sinon.stub(fs, "readFileSync");
+    });
+
+    beforeEach(() => {
+        stubSafelyReadJsonFile.withArgs(path.normalize(`${testRepositoryPath}/.openpublishing.publish.config.json`))
+            .returns({
+                docsets_to_publish: [
+                    {
+                        docset_name: "docset1",
+                        build_source_folder: "sourceFolder1"
+                    }
+                ]
+            });
+        testEventBus.clear();
+    });
+
+    after(() => {
+        sinon.restore();
+        testEventBus.dispose();
+    });
+
+    describe("No validation result", () => {
+        before(() => {
+            stubFsExistsSync
+                .withArgs(path.normalize(`${testRepositoryPath}/_site/sourceFolder1/.errors.log`))
+                .returns(false);
+        });
+
+        beforeEach(() => {
+            testEventBus.clear();

```

Comment:
consider avoid the nested `describe`, the outer `beforeEach` is ignored here

Reply:
my fault, seems the outer `beforeEach` will also be applied for the insider it. and those several `it` are testing the different cases of the same pre-condition, I prefer to put them into one `describe`. Agree?

---
In reply to: [394072812](https://github.com/docascode/vscode-docs-build/pull/94#discussion_r394072812) [](ancestors = 394072812)